File: part1/hs/ex3a.hs
quicksort :: (Ord a) => [a] -> [a]
quicksort [] = []
quicksort (x:xs) =
    let left = quicksort [ a | a <- xs, a <= x ]
        right = quicksort [ a | a <- xs, a > x ]
    in left ++ [x] ++ right

binarysearch :: (Ord a, Eq a) => a -> [a] -> Maybe a
binarysearch _ [] = Nothing
binarysearch x xs
    | pivot == x = Just x
    | pivot > x = binarysearch x left
    | otherwise = binarysearch x right
    where pivot = xs !! ((length xs) `div` 2)
          left = [ xs !! a | a <- [0,1..((length xs) `div` 2) - 1] ]
          right = [ xs !! a | a <- [((length xs) `div` 2) + 1 .. (length xs) - 1] ]



search :: (Ord a, Eq a) => [a] -> a -> Maybe a
search xs x = binarysearch x $ quicksort xs

File: part1/js/ex3a.js
/*
 * Implement the BinarySearch algorithm using a list.
 *
 */

// (Ord a) => [a] -> [a]
function quicksort(list) {
    if(list.length === 0) {
        return [];
    }

    var pivot = list.shift(),
        left = list.filter(function (a) {
            return a <= pivot
        }),
        right = list.filter(function (a) {
            return a > pivot
        });

    return quicksort(left).concat(pivot).concat(quicksort(right));
}

// Assumes list is sorted
// (Eq a) => [a] -> a -> Maybe a
function binarySearch(list, item) {
    var mid = Math.floor(list.length / 2),
        slice;

    if(mid === 0) {
        return (item === list[mid] ? item : null);
    }

    if(list[mid] === item) {
        return item;
    }

    if(list[mid] > item) {
        return search(list.slice(0, mid), item);
    } else {
        return search(list.slice(mid + 1, list.length), item);
    }
}


// (Eq a, Ord a) => [a] -> a -> Maybe a
function search(list, item) {
    list = quicksort(list);
    return binarySearch(list, item);
}

module.exports = search;

// Testing module

var testcase = [5,4,3,5,6,2,1,7];
console.log(search(testcase, 2));
